### PR TITLE
rpi3: use upstream ATF and U-boot

### PR DESCRIPTION
**Pull request by igoropaniuk**
Depends on :
https://github.com/OP-TEE/optee_os/pull/2394
https://github.com/OP-TEE/manifest/pull/119

`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

### CODE REVIEW

**Comment by igoropaniuk** [comment]
FYI, work is still in progress

@grandpaul
BL1+BL2+BL31+BL32 (OP-TEE) are booting OK, but currently I'm facing issues when control is passed to BL33 U-boot.

Did you manage to bring up setup with U-boot, or you just followed instructions in https://github.com/ARM-software/arm-trusted-firmware/blob/master/docs/plat/rpi3.rst and used TF Bootstrap instead (https://github.com/AntonioND/rpi3-arm-tf-bootstrap)?

**Comment by igoropaniuk** [comment]
```
D/TC:0 tee_ta_register_ta_store:534 Registering TA store: 'REE' (priority 10)
D/TC:0 tee_ta_register_ta_store:534 Registering TA store: 'Secure Storage TA' (priority 9)
D/TC:0 mobj_mapped_shm_init:559 Shared memory address range: 11200000, 13200000
I/TC:  Initialized
D/TC:0 init_primary_helper:917 Primary CPU switching to normal world boot
ERROR:   Interrupt not expected to be routed to EL3PANIC in EL3 at x30 = 0x00000000100e40e0
x0 =		0x0000000000000000
x1 =		0x000000003f215040
x2 =		0x0000000000000060
x3 =		0x0000000000000000
x4 =		0x0000000000000001
x5 =		0x0000000000000000
x6 =		0x00000000100ecf90
x7 =		0x0000000000000020
x8 =		0x0000000000000064
x9 =		0x000000001017e640
x10 =		0x0000000000000060
x11 =		0x00000000100e7f30
x12 =		0x0000000000000033
x13 =		0x0000000000000033
x14 =		0x0000000000000000
x15 =		0x00000000100e3fa0
x16 =		0x00000000600003c4
x17 =		0x00000000101000f8
x18 =		0x0000000000000e38
x19 =		0x0000000000000000
x20 =		0x0000000000000000
x21 =		0x0000000000000000
x22 =		0x0000000010106120
x23 =		0x0000000000000000
x24 =		0x0000000000000600
x25 =		0x00000000100ecf90
x26 =		0x00000000100ecf70
x27 =		0x0000000000000001
x28 =		0x0000000000000020
x29 =		0x00000000100e8f10
scr_el3 =		0x0000000000000e38
sctlr_el3 =		0x0000000030cd183f
cptr_el3 =		0x0000000000000000
tcr_el3 =		0x0000000080803520
daif =		0x00000000000002c0
mair_el3 =		0x00000000004404ff
spsr_el3 =		0x00000000600003c4
elr_el3 =		0x00000000101000f8
ttbr0_el3 =		0x00000000100ecac0
esr_el3 =		0x000000005e000000
far_el3 =		0x00351e01f10d9dac
spsr_el1 =		0x0000000000000000
elr_el1 =		0x0000000000000000
spsr_abt =		0x0000000000000000
spsr_und =		0x0000000000000000
spsr_irq =		0x0000000000000000
spsr_fiq =		0x0000000000000000
sctlr_el1 =		0x0000000030d8180f
actlr_el1 =		0x0000000000000000
cpacr_el1 =		0x0000000000000000
csselr_el1 =		0x0000000000000000
sp_el1 =		0x0000000010157ee0
esr_el1 =		0x0000000000000000
ttbr0_el1 =		0x000000001017e000
ttbr1_el1 =		0x0000000000000000
mair_el1 =		0x000000000000ff04
amair_el1 =		0x0000000000000000
tcr_el1 =		0x0000000080803f20
tpidr_el1 =		0x0000000000000000
tpidr_el0 =		0x0000000000000000
tpidrro_el0 =		0x0000000000000000
dacr32_el2 =		0x0000000000000000
ifsr32_el2 =		0x0000000000000000
par_el1 =		0xff00000010100980
mpidr_el1 =		0x0000000080000000
afsr0_el1 =		0x0000000000000000
afsr1_el1 =		0x0000000000000000
contextidr_el1 =		0x0000000000000000
vbar_el1 =		0x0000000010100800
cntp_ctl_el0 =		0x0000000000000002
cntp_cval_el0 =		0x7bf0ddb7bde9d4ff
cntv_ctl_el0 =		0x0000000000000000
cntv_cval_el0 =		0x34223a51c96c1a40
cntkctl_el1 =		0x0000000000000000
sp_el0 =		0x00000000100e8f10
isr_el1 =		0x0000000000000000
cpuectlr_el1 =		0x0000000000000040
cpumerrsr_el1 =		0x0000000000000000
l2merrsr_el1 =		0x0000000000000000
cpuactlr_el1 =		0x00001000090ca000
```

**Comment by grandpaul** [comment]
As a log here. When building u-boot, we have to config CONFIG_SYS_TEXT_BASE=0x10300000 and use it as a BL33 payload.

**Comment by grandpaul** [comment]
Hi. I think I need to clarify this a bit.
In build's ATF [1], there's an empty plat_interrupt_type_to_line() in plat_helpers.S. 
It is bad for upstream because it is supposed to handle the interrupt there. Thus in the upstream ATF, the same function plat_interrupt_type_to_line() is put to rpi3_common.c and calling panic() to make sure that we really don't need to handle the interrupt because there's no GIC for RPi3.

However, when tracing the code, I found opteed in ATF did call that function. I think it is some initialization but I'm not sure. The path is register_interrupt_type_handler() -> set_routing_model() -> set_scr_el3_from_rm() -> plat_interrupt_type_to_line()

It seems we need to figure out what does plat_interrupt_type_to_line() really do and implement it correctly in ATF.

[1] https://github.com/OP-TEE/build/blob/master/docs/rpi3.md

**Comment by antonio-nino-diaz-arm** [comment]
Ok, then it is needed to setup the interrupt routing correctly.

The RPi3 doesn't have secure devices, so I'd say that all the interrupts have to be routed to the normal world to be handled by Linux, right?

**Comment by antonio-nino-diaz-arm** [comment]
If this is an acceptable behaviour I can create a patch to fix that in the TF.

**Comment by antonio-nino-diaz-arm** [comment]
Can you tell me if this works for you, @grandpaul? https://github.com/ARM-software/arm-trusted-firmware/pull/1464

File: rpi3.mk [hunk 1]
```diff
@@ -57,22 +58,23 @@ ARM_TF_EXPORTS ?= \
...
-	cd $(ARM_TF_OUT) && \
-	  dd if=/dev/zero of=scratch bs=1c count=131072 && \
-	  cat $(ARM_TF_BIN) scratch > $(ARM_TF_TMP) && \
-	  dd if=$(ARM_TF_TMP) of=$(ARM_TF_HEAD) bs=1c count=131072 && \
-	  cat $(ARM_TF_HEAD) $(OPTEE_OS_PAGER) > $(ARM_TF_BOOT) && \
-	  rm scratch $(ARM_TF_TMP) $(ARM_TF_HEAD)
+arm-tf: optee-os u-boot
+	$(ARM_TF_EXPORTS) $(MAKE) -C $(ARM_TF_PATH) $(ARM_TF_FLAGS) all fip
+	@cp $(ARM_TF_OUT)/bl1.bin $(ARM_TF_OUT)/bl1.pad.bin
+	@truncate --size=65536 $(ARM_TF_OUT)/bl1.pad.bin
```
Comment (antonio-nino-diaz-arm) [inline rpi3.mk]: This is now 131072 instead of 65536.

Comment (antonio-nino-diaz-arm) [reply rpi3.mk]: Actually, take a look at this PR: https://github.com/ARM-software/arm-trusted-firmware/pull/1478 Maybe we can get rid of this completely.

**Comment by igoropaniuk** [comment]
Tested with upstream ATF + U-boot as BL33 (with changed `CONFIG_SYS_TEXT_BASE` to `0x11000000`) + Raspberry Linux 4.14 fork (https://github.com/raspberrypi/linux/tree/rpi-4.14.y + with proper optee node in dts + disabled FIQ in dwc_otg). Everything is a built for ARM64.

```
$ xtest
+-----------------------------------------------------
16081 subtests of which 0 failed
74 test cases of which 0 failed
0 test case was skipped
TEE test application done!
```

Will create PR in manifest.git + linux branch in https://github.com/linaro-swg/linux.git based on Raspberry Linux 4.14 (https://github.com/raspberrypi/linux/tree/rpi-4.14.y + with proper optee node in dts) soon.

**Comment by igoropaniuk** [comment]
Thanks a lot @grandpaul and @antonio-nino-diaz-arm for your contribution/help!

**Comment by igoropaniuk** [comment]
@jbech-linaro I've updated rpi3.md and enabled NFS boot (let's switch to RST format in another PR).
Please feel free to test and leave your T-b tags. Thanks!

**Comment by igoropaniuk** [comment]
@antonio-nino-diaz-arm just noticed that you've approved this PR.
Could you please explicitly leave R-b tag here, so I can attach it to my commit message?
Thanks

**Comment by antonio-nino-diaz-arm** [comment]
I've clicked approve because I've just noticed that this PR still had a "request changes" from me from a few weeks ago, and I didn't want to create confusion. But sure:
```
Reviewed-by: Antonio Nino Diaz <antonio.ninodiaz@arm.com>
```

**Comment by igoropaniuk** [comment]
Squashed commits, applied R-b tag and rebased on the latest master, ready for merge. Thanks!

**Comment by igoropaniuk** [comment]
Rebased on the latest master

**Pull request merged**